packages/vega-parser/src/parsers/guides/axis-config.js

```diff
@@ -1,14 +1,57 @@
 import {Bottom, Top} from './constants';
 import {extend} from 'vega-util';
+import { isSignal } from '../../util';
+import { allAxisOrientSignalRef, xyAxisSignalRef } from './axis-util';
 
 export default function(spec, scope) {
   var config = scope.config,
       orient = spec.orient,
-      xy = (orient === Top || orient === Bottom) ? config.axisX : config.axisY,
-      or = config['axis' + orient[0].toUpperCase() + orient.slice(1)],
-      band = scope.scaleType(spec.scale) === 'band' && config.axisBand;
+      band = scope.scaleType(spec.scale) === 'band' && config.axisBand,
+      xy,
+      or;
 
-  return (xy || or || band)
+  if (isSignal(spec.orient)) {
+    var axisX = config.axisX || {},
+        axisY = config.axisY || {},
+        axisTop = config.axisTop || {},
+        axisBottom = config.axisBottom || {},
+        axisLeft = config.axisLeft || {},
+        axisRight = config.axisRight || {},
+        axisXYConfigKeys = new Set(
+          Object.keys(axisX)
+            .concat(axisY)
+        ),
+        axisOrientConfigKeys = new Set(
+          Object.keys(axisTop)
+            .concat(Object.keys(axisBottom)
+            .concat(Object.keys(axisLeft)
+            .concat(Object.keys(axisRight))))
+        );
+
+
+    xy = {};
+    for (var prop of axisXYConfigKeys) {
+      xy[prop] = xyAxisSignalRef('x', spec.orient.signal, axisX[prop], axisY[prop]);
+    }
+
+    or = {};
+    for (prop of axisOrientConfigKeys) {
+      or[prop] = allAxisOrientSignalRef(
+        spec.orient.signal,
+        axisTop[prop],
+        axisBottom[prop],
+        axisLeft[prop],
+        axisRight[prop]
```

Comment:
@jheer If a prop exists in one axis config but not the others, this will conditionally set that property to `null`. Perhaps for limited use cases this okay (the user would need to match props in the configs if they intend to swap axis orient), but ideally this would instead conditionally set it to the default value. Is there a readily available approach to implementing this?

Reply:
Perhaps use the top-level config.axis as backup? Though that might still write null values over lower-level mark or style config entries. You are learning the hard way why I did not make axis orient a signal in the past...

Reply:
It is quite a hairy refactor...

Ah, I hadn't thought about falling back to the top-level config. I also looked into the default style config (in `vega-parser/src/config.js`) and implemented a fallback routine that accesses those values. I believe this solves the issue—it seems like `null` for properties that are not in the axis config or the default styles are handled gracefully by the renderer (correct me if I'm wrong).